---
title: 前端异常监控
date: 2017-12-10 12:38:10
tags:
---

如果一个系统不做异常监控，那么上线后完全不能感知到系统的异常。一个好的系统不应该让用户来反馈异常才找到自己的缺陷，异常上报对一个能自我容错的系统起到至关重要的作用

要做异常上报之前，先要对系统中经常出现的异常做一次归纳。我将错误类型分成两种，一种是编译错误，另一种是执行错误。顾名思义，__编译错误__是检查js语法时候报错，__执行错误__是执行js代码时候报错。

<br />
### 编译报错
```javascript
function test (a, b) {};
test(10 20);
// Uncaught SyntaxError: missing ) after argument list
```
编译报错的类型有:
语法错误`SyntaxError`
当引擎检查js语法，发现异常，会抛出错误，'同个块'的js都不会执行。SyntaxError不能被try-catch捕捉。
> 同个块指的是在相同script标签内、同一个js外链里的代码
<!--more-->

<br />
### 执行报错
```javascript
new Array(-20); // Uncaught RangeError: Invalid array length

var arr = {};arr.split(''); // Uncaught TypeError: arr.split is not a function

throw new EvalError('eval'); // Uncaught EvalError: eval

decodeURI('%0'); // Uncaught URIError: URI malformed

a = b; // Uncaught ReferenceError: b is not defined

// net::ERR_FILE_NOT_FOUND || 404 (Not Found)
var img = new Image();
img.src = 'noExist.jpg';
document.body.append(img);

// No 'Access-Control-Allow-Origin' header is present on the requested resource. Origin '***' is therefore not allowed access. The response had HTTP status code 404.
var xmlhttprequest = new XMLHttpRequest();
xmlhttprequest.open("GET", 'http://localhost:8080/index.html', true);
xmlhttprequest.send();
```
数值超出`RangeError`
类型错误`TypeError`
eval错误`EvalError`
url错误`URIError`
无定义错误`ReferenceError`
文件不存在`ERR_FILE_NOT_FOUND || 404 (Not Found)`
同源政策`No 'Access-Control-Allow-Origin' header`
执行到相应js时报错，发现异常，'同个块'内中断单步执行。可以被try-catch捕捉到错误。
> 文件找不到，样式、脚本外链报的是net::ERR_FILE_NOT_FOUND，html、图片报的是404 (Not Found)


<br />
### 所有错误类型归纳出来后，就可以考虑捕捉错误的方案了。
* try-catch
擅长处理可控的异常，代码混淆后获取信息比window.onerror更准确，但是不能捕捉异步异常, 不能捕捉编译错误.
不能捕捉编译错误上面已经有提到，不能捕捉异步异常，例如：
```javascript
try {
    setTimeout(function () {
        throw new Error('asyn');
    }, 0)
} catch (e) {
    console.log(e); // catch捕捉不到error的异常
}
```

* window.onerror
可以捕捉不可控的异常，可以捕捉编译异常、数值超出错误、类型错误、异步操作报错等异常，但是不能捕捉iframe的错误、不能捕捉文件不存在的错误.
```javascript
var tempOnerror = window.onerror;
window.onerror = function (msg, url, l) {
    // TODO
    // ...

    tempOnerror.apply(window, arguments);

    return true; // 禁止错误打印到console
}
```

* window监听error事件
可以捕捉文件不存在的错误、捕捉跨域限制的错误.但是捕捉的请求错误不能获取到url等等信息
```javascript
window.addEventListener('error', function (error) {
    console.log(error);
}, true); // 事件capturing阶段捕捉错误，file.onerror事件不冒泡
```

* 节点onerror属性
当相应节点报错后，直接捕捉异常信息
```javascript
<img src="noExist.jpg" onerror="errorFn(this)">

var img = new Image(); // net::ERR_FILE_NOT_FOUND 文件没有找到错误无法捕捉
img.onerror = function (e) {
    console.log(e);
};
img.src = 'noExist.jpg';
document.body.append(img);

var xmlhttprequest = new XMLHttpRequest();
xmlhttprequest.open("GET", 'http://localhost:8080/index.html', true);
xmlhttprequest.onerror = function (e) {
    // e.preventDefault();
}
xmlhttprequest.send();
```

<br />
### 比较几种方案后，最后我选择window.onerror + window监听error事件 + 节点onerror属性相结合
`window.onerror`可以对不可控的异常捕捉，`window监听error事件`可以捕捉文件不存在的异常, `节点onerror属性`可以捕捉请求的异常，这种就可以收集请求url、data等等的信息。


<br />
### 封装成插件
...